elfo-logger/src/line_buffer.rs

```diff
@@ -0,0 +1,274 @@
+use std::mem;
+
+use crate::line_transaction::Line;
+
+// BufferConfiguration
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct BufferConfig {
+    pub(crate) max_line_size: usize,
+}
+
+impl Default for BufferConfig {
+    fn default() -> Self {
+        Self {
+            max_line_size: usize::MAX,
+        }
+    }
+}
+
+// Repr
+
+#[derive(Debug)]
+struct Repr<'a> {
+    buf: &'a mut LineBuffer,
+
+    // We don't wanna write unfinished data, so there must be a way to
+    // revert changes made by `CurrentLine`
+    pre_start_buffer_size: usize,
+}
+
+// CurrentLine
+
+#[derive(Debug)]
+pub(crate) struct CurrentLine<'a>(Repr<'a>);
+
+impl<'a> Line for CurrentLine<'a> {
+    fn try_commit(self) -> bool {
+        true
+    }
+
+    fn total_bytes(&self) -> usize {
+        self.len()
+    }
+
+    fn meta_mut(&mut self) -> &mut String {
+        &mut self.0.buf.buffer
+    }
+
+    fn payload_mut(&mut self) -> &mut String {
+        &mut self.0.buf.payload
+    }
+
+    fn fields_mut(&mut self) -> &mut String {
+        &mut self.0.buf.fields
+    }
+}
+
+impl<'a> CurrentLine<'a> {
+    fn meta_len(&self) -> usize {
+        self.0.buf.buffer.len() - self.0.pre_start_buffer_size
+    }
+
+    fn len(&self) -> usize {
+        self.meta_len() + self.0.buf.payload.len() + self.0.buf.fields.len()
+    }
+}
+
+impl<'a> CurrentLine<'a> {
+    fn probe_size_limit(&mut self) {
+        let len = self.len();
+        let mut need_to_erase = if len > self.0.buf.cfg.max_line_size {
+            len - self.0.buf.cfg.max_line_size
+        } else {
+            return;
+        };
+
+        let payload_len = self.0.buf.payload.len();
+        let fields_len = self.0.buf.fields.len();
+        let meta_len = self.meta_len();
+
+        let payload_part = payload_len.min(need_to_erase);
+        need_to_erase -= payload_part;
+        let fields_part = fields_len.min(need_to_erase);
+        need_to_erase -= fields_part;
+        let meta_part = meta_len.min(need_to_erase);
+
+        self.0.buf.payload.truncate(payload_len - payload_part);
+        self.0.buf.fields.truncate(fields_len - fields_part);
+        self.0
+            .buf
+            .buffer
+            .truncate(self.0.buf.buffer.len() - meta_part);
```

Comment:
What about adding some info that the line is truncated? For instance, in the dumper, we add " TRUNCATED"